zerver/lib/actions.py

```diff
@@ -2226,20 +2226,20 @@ def do_add_reaction(
     emoji_code: str,
     reaction_type: str,
 ) -> None:
+    """Should be called while holding a SELECT FOR UPDATE lock
+    (e.g. via access_message(..., lock_message=True)) on the
+    Message row, to prevent race conditions.
+    """
+
     reaction = Reaction(
         user_profile=user_profile,
         message=message,
         emoji_name=emoji_name,
         emoji_code=emoji_code,
         reaction_type=reaction_type,
     )
-    try:
-        reaction.save()
-    except django.db.utils.IntegrityError:  # nocoverage
-        # This can happen when a race results in the check in views
-        # code not catching an attempt to double-add a reaction, or
-        # perhaps if the emoji_name/emoji_code mapping is busted.
-        raise JsonableError(_("Reaction already exists."))
```

Comment:
Can you tweak the commit message to explain why we get to delete this a bit more explicitly?  I think the story is "... because `lock_message` will wait until the duplicate transaction finishes".

"""
As a nice side-effect, we also eliminate the previous race condition from double
adding reactions (by waiting for the message row lock), and thus get to delete 
somewhat messy `IntegrityError` error handling logic.
"""

Reply:
Done.